### ISSUE TITLE
azurerm_function_app os_type only works if it's 'linux'.  empty string  Error: expected os_type to be one of [linux], got

### ISSUE THREAD

**Issue by marc-gmail-towersap** [terraform-providers/terraform-provider-azurerm#11859]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.15.4
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.60.0

### Affected Resource(s)

* `azurerm_function_app`

### Terraform Configuration Files

```hcl
variable "prefix" {
    description = "A prefix used for all resources"
    type = string 
    default = "blah-funct"
}
variable "environment" {
    description = "is it dev, qa, UAT, prod, or other?"
    type = string
    default = "dev"
}
variable "location" {
    description = "The Azure Region in which all resources should be provisioned"
    type = string
    default = "East US"
}
 variable "asp_kind" {
    description = "App Service Plan kind"
    type = string
    default = "Windows"
} 
locals {
    ## osType for the azure function is either linux or "" (windows)
    osType = var.asp_kind == "Linux" ? "linux" : ""
    saPrefix = lower(replace(var.prefix, "-", "") )
}
resource "azurerm_resource_group" "thisrg" {
  name     = "${var.prefix}-${var.environment}-rg"
  location = var.location
  // tags     = local.tags
  # count    = var.create_or_use_rg == "create" ? 1 : 0
} 
resource "azurerm_app_service_plan" "asp" {
  name                = "${var.prefix}-${var.environment}-asp"
  location            = var.location
  resource_group_name = azurerm_resource_group.thisrg.name
  kind                = var.asp_kind
  reserved            = true
  
  sku {
    tier = "Basic"
    size = "B1"
  }
}

resource "azurerm_storage_account" "sa" {
  name                = "${local.saPrefix}${var.environment}sa"
  resource_group_name = azurerm_resource_group.thisrg.name
  location            = var.location
  account_tier        = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_function_app" "fa" {
  name                = "${var.prefix}-${var.environment}-function"
  location            = var.location
  // resource_group_name = local.rgname
  resource_group_name = azurerm_resource_group.thisrg.name
  app_service_plan_id = azurerm_app_service_plan.asp.id
  storage_account_name = azurerm_storage_account.sa.name
  storage_account_access_key = azurerm_storage_account.sa.primary_access_key
  os_type             = local.osType
  // depends_on = [data.azurerm_resource_group.thisrg]
}

```

### Expected Behaviour

 + resource "azurerm_function_app" "fa" {
      + app_service_plan_id            = (known after apply)
      + app_settings                   = (known after apply)
      + client_affinity_enabled        = (known after apply)
      + custom_domain_verification_id  = (known after apply)
      + default_hostname               = (known after apply)
      + enable_builtin_logging         = true
      + enabled                        = true
      + https_only                     = false
      + id                             = (known after apply)
      + kind                           = (known after apply)
      + location                       = "eastus"
      + name                           = "dataverse-funct-marcdev-function"
      + os_type                        = ""
      + outbound_ip_addresses          = (known after apply)
      + possible_outbound_ip_addresses = (known after apply)
      + resource_group_name            = "dataverse-funct-marcdev-rg"
      + site_credential                = (known after apply)
      + storage_account_access_key     = (sensitive value)
      + storage_account_name           = "dataversefunctmarcdevsa"
      + storage_connection_string      = (sensitive value)
      + version                        = "~1"

### Actual Behaviour
```hcl
PS C:\git_repos\blah\Terraform\Templates\AzureFunction> terraform plan             
╷
│ Error: expected os_type to be one of [linux], got
│
│   with azurerm_function_app.fa,
│   on function.tf line 22, in resource "azurerm_function_app" "fa":
│   22: resource "azurerm_function_app" "fa" {
│
╵
```

### Steps to Reproduce

See above for tf code.
1. `terraform plan`
2. it works if I run 'terraform plan -var asp_kind=Linux
3. Documentation says (I added the bold), 
os_type - (Optional) A string indicating the Operating System type for this function app.

NOTE:
This value will be linux for Linux derivatives, or **an empty string for Windows (default)**. When set to linux you must also set azurerm_app_service_plan arguments as kind = "Linux" and reserved = true

### References

* #0000

**Comment by eltimmo** [comment]
I'm having similar issues. I've added both ```os_type``` and ```client_cert_mode``` attributes  in a module with null'd variables by default. When doing a plan \ apply terraform wants to set these to = "", however I just don't want to set these at all nor should there be any change as these have never been set.

![image](https://user-images.githubusercontent.com/27624232/122180323-3b3e3380-ce80-11eb-816e-51e72eca31a4.png)

**Comment by marc-gmail-towersap** [comment]
I got around my problem by doing two different definitions, a winfa and a linuxfa (instead of a generic fa as shown above), and doing a count = local.isLinux ? 1 : 0  (which then builds the linuxfa block) and a count = local.isLinux ? 0 : 1 for winfa.
If isLinux is true.  eltimmo, you can probably do the same trick.

**Comment by eltimmo** [comment]
Unless I'm mistaken there's something wrong with these how these implemented in the provider or the documentation?. These are both stated as being optional. However when these are not present, or set to null via a variable terraform still tries to set these to = "".  Also the only valid option for ```os_type``` is ```linux```, terraform appears to want to set this to ```""``` but this is never set to I get prompted even after multiple applies.

**client_cert_mode** - (Optional) The mode of the Function App's client certificates requirement for incoming requests. Possible values are Required and Optional.

**os_type** - (Optional) A string indicating the Operating System type for this function app.

**Comment by JeremyBouchet** [comment]
Hello,
I have an **Azure Function with Windows OS** and I encounter also the error "**Error: expected os_type to be one of [linux], got" when I executes "terraform plan**".

My Terraform code :

```hcl
resource "azurerm_function_app" "AF-Core-generic" {
  name                        = "AF-${var.client}-${var.environment}-${var.project_Flux}-IT-CORE-generic"
  location                    = module.resource_group_CORE.out_rg_location
  resource_group_name         = module.resource_group_CORE.out_rg_name
  app_service_plan_id         = azurerm_app_service_plan.ASP-Core.id
  https_only                  = "true"
  version                     = "~2"
  storage_account_access_key  = module.storage_account_CORE.out_storage_primary_connection_string
  os_type                     = ""
  enable_builtin_logging      = "false"
  tags                        = merge(var.default_tags, var.default_Core_tags, var.AF-Core-generic_tags)
```
But in the Terraform documentation it's explained that the os_type variable must be "_an empty string for Windows (default)._"

Also I tried with the "lifecycle / ignore_changes" parameters to ignore the "os_type" variable but I have always the error.

Is anyone had find a solution to fix this error ?

**Comment by tombuildsstuff** [comment]
Fixed via #12482
